src/gui/qgsfieldformwidget.h

```diff
@@ -0,0 +1,256 @@
+/***************************************************************************
+   qgsfieldformwidget.h
+    --------------------------------------
+    begin                : September 2020
+    copyright            : (C) 2020 by Ivan Ivanov
+    email                : ivan@opengis.ch
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************/
+
+#ifndef QGSFIELDFORMWIDGET_H
+#define QGSFIELDFORMWIDGET_H
+
+#include "ui_qgsfieldformwidget.h"
+#include "qgsguiutils.h"
+#include "qgsapplication.h"
+
+#include "qgis_gui.h"
+#include "qgsfield.h"
+
+#include <QRegExpValidator>
+#include <QRegExp>
+
+class QgsVectorLayer;
+
+/**
+* \ingroup gui
+* Dialog to set up parameters to create a new field in a vector layer, and on accept() to create it and add it to the layers
+* \note unstable API (will likely change)
+* \since QGIS 3.16
+*/
+class GUI_EXPORT QgsFieldFormWidget : public QWidget, private Ui::QgsFieldFormWidgetBase
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+     * List of advanced fields to be shown. If `Neither` is chosen, the whole advanced fields UI is hidden.
+     */
+    enum class AdvancedField
+    {
+      // sip complains if it is called `None`
+      Neither = 1 << 0,
+      Comment = 1 << 1,
+      Alias = 1 << 2,
+      IsNullable = 1 << 3,
+      All = Comment | Alias | IsNullable,
+    };
+
+    Q_DECLARE_FLAGS( AdvancedFields, AdvancedField )
+    Q_FLAG( AdvancedFields )
+
+    //! Constructor
+    QgsFieldFormWidget( AdvancedFields advancedFields = QgsFieldFormWidget::AdvancedField::Neither, QWidget *parent SIP_TRANSFERTHIS = nullptr );
+
+    //! Adds string field type to given field form with default settings
+    static void addStringType( QgsFieldFormWidget *fieldForm )
```

Comment:
I'm not a fan of this API -- I think a better approach would be to remove all the specific add* methods and just have 

    void setSupportedTypes( const QList< QVariant::Type >& types );

instead

Reply:
What about the typename here?

Reply:
Type name is completely optional. It's safe to set it to an empty string without any consequence.